packages/sdk/src/filter_next/filter.ts

```diff
@@ -0,0 +1,189 @@
+import { Logger } from "@waku/utils";
+
+import { PeerManager } from "../peer_manager/index.js";
+
+import { Subscription } from "./subscription.js";
+import { FilterConstructorParams } from "./types.js";
+
+const log = new Logger("sdk:next-filter");
+
+type PubsubTopic = string;
+
+export class Filter implements IFilter {
+  private readonly libp2p: Libp2p;
+  private readonly protocol: FilterCore;
+  private readonly peerManager: PeerManager;
+  private readonly connectionManager: ConnectionManager;
+
+  private readonly config: FilterOptions;
+  private subscriptions = new Map<PubsubTopic, Subscription>();
+
+  public constructor(params: FilterConstructorParams) {
+    this.config = {
+      numPeersToUse: 2,
+      pingsBeforePeerRenewed: 3,
+      keepAliveIntervalMs: 60_000,
+      ...params.options
+    };
+
+    this.libp2p = params.libp2p;
+    this.peerManager = params.peerManager;
+    this.connectionManager = params.connectionManager;
+
+    this.protocol = new FilterCore(
+      this.onIncomingMessage.bind(this),
+      params.connectionManager.pubsubTopics,
+      params.libp2p
+    );
+  }
+
+  public async subscribe<T extends IDecodedMessage>(
+    decoder: IDecoder<T> | IDecoder<T>[],
+    callback: Callback<T>
+  ): Promise<boolean> {
+    const decoders = Array.isArray(decoder) ? decoder : [decoder];
+
+    if (decoders.length === 0) {
+      throw Error("Cannot subscribe with 0 decoders.");
+    }
+
+    const pubsubTopics = decoders.map((v) => v.pubsubTopic);
+    const contentTopics = decoders.map((v) => v.contentTopic);
+
+    // doing this for simplicity, we can enable subscription for more than one PubsubTopic at once later when requested
+    if (!this.isSamePubsubTopic(decoders)) {
+      throw Error(
+        `Cannot subscribe to more than one pubsub topic at the same time, got pubsubTopics:${pubsubTopics}`
+      );
+    }
+
+    log.info(
+      `Subscribing to content topic: ${contentTopics}, pubsub topic: ${pubsubTopics}`
+    );
+
+    const supportedPubsubTopic = this.connectionManager.pubsubTopics.includes(
+      pubsubTopics[0]
+    );
+
+    if (!supportedPubsubTopic) {
+      throw Error(
+        `Pubsub topic ${pubsubTopics[0]} has not been configured on this instance.`
+      );
+    }
+
+    let subscription = this.subscriptions.get(pubsubTopics[0]);
+    if (!subscription) {
+      subscription = new Subscription({
+        pubsubTopic: pubsubTopics[0],
+        libp2p: this.libp2p,
+        protocol: this.protocol,
+        config: this.config,
+        peerManager: this.peerManager
+      });
+      subscription.start();
+    }
+
+    const result = await subscription.add(decoders, callback);
+    this.subscriptions.set(pubsubTopics[0], subscription);
+
+    log.info(
+      `Subscription ${result ? "successful" : "failed"} for content topic: ${contentTopics}`
+    );
+
+    return result;
+  }
+
+  public async unsubscribe<T extends IDecodedMessage>(
+    decoder: IDecoder<T> | IDecoder<T>[]
```

Comment:
probably out of scope for this PR, but it seems strange that we have to pass the decoder object(s) in order to unsubscribe

Reply:
that is something inherited from how Filter works and how we try to make it work
- how it works: req-res protocol for subscribing 
- how we make it work: event-like API for subscribing via a callback 

because we are trying to proxy one into another we get this weir feeling like maybe better to use `contentTopics` directly 

but I think we shouldn't go this way and continue providing JS native API experience and that is Event driven API